src/marqo/core/distributed_lock/abstract_distributed_lock.py

```diff
@@ -0,0 +1,61 @@
+import threading
+from abc import ABC, abstractmethod
+
+from kazoo.recipe.lock import Lock
+
+from marqo.core.distributed_lock.marqo_kazoo_client import MarqoKazooClient
+from marqo.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+class AbstractExpiringDistributedLock(ABC):
+    """Abstract class for a distributed lock with expiration.
+
+    This lock is used to ensure that a resource is not accessed concurrently by multiple processes. It must also have a
+    watchdog that releases the lock if it is held for too long.
+    The lock should also have __enter__ and __exit__ methods to be used as a context manager.
+    """
+
+    @abstractmethod
+    def __init__(self, zookeeper_client: MarqoKazooClient, path: str, max_lock_period: float, watchdog_interval: float):
+        """Initialize the distributed lock.
+        Args:
+            zookeeper_client: The Zookeeper client.
+            path: The lock path.
+            max_lock_period: The maximum period of time the lock can be held.
+            watchdog_interval: The interval at which the watchdog checks the lock status.
+        """
+        self.zookeeper_client = zookeeper_client
```

Comment:
I think this abstraction is a bit questionable. Lots of Zookeeper code is leaking into the abstraction so it's not independent of Zookeeper. Also see my comments on the implementation class below